components/list/list.js

```diff
@@ -64,13 +71,65 @@ class List extends LitElement {
 		return [layout, specialCases];
 	}
 
+	firstUpdated(changedProperties) {
+		super.firstUpdated(changedProperties);
+		this._checked = {};
+		this.addEventListener('d2l-list-item-checked', this._onlistItemChecked.bind(this));
+	}
+
+	itemsChecked() {
+		return Object.keys(this._checked).filter(ref => this._checked[ref]);
+	}
+
 	render() {
 		return html`
 			<div role="list" class="d2l-list-container">
 				<slot></slot>
 			</div>
 		`;
 	}
+
+	selectionState() {
+		const checkedListItems = this.itemsChecked();
+		if (!checkedListItems || checkedListItems.length < 1) {
+			return selectableListStates.none;
+		}
+
+		const uncheckedListItems = this.querySelectorAll('d2l-list-item:not([checked])');
+		if (uncheckedListItems.length < 1) {
+			return selectableListStates.all;
+		}
+
+		return selectableListStates.indeterminate;
+	}
+
+	selectAll() {
+		if (!this.selectable) {
+			return;
+		}
+		const uncheckedListItems = this.querySelectorAll('d2l-list-item:not([checked])');
+		if (uncheckedListItems.length < 1) {
+			const checkedListItems = this.querySelectorAll('d2l-list-item[checked]');
+			checkedListItems.forEach(listItem => {
+				listItem.toggleAttribute('checked');
+			});
+		} else {
+			uncheckedListItems.forEach(listItem => {
+				listItem.toggleAttribute('checked');
+			});
+		}
+	}
+
+	_onlistItemChecked(event) {
+		event.stopPropagation();
+
+		this._checked[event.detail.ref] = event.detail.checked;
+		this.dispatchEvent(new CustomEvent('change', {
```

Comment:
The name `change` is a native event name used on form inputs. I don't think we should reuse it here. I'd suggest `d2l-list-selection-change`.

Reply:
I was picturing the use case to be `<d2l-list @change="${this._onListChange">` With this it will be `<d2l-list @d2l-list-selection-change="${this._onListChange">` which is longer but still work. Thoughts?

Reply:
Anything inside the list that bubbles up a `change` event could cause this logic to run.  I think it's possible to overload the native `change`, but I think it's safer and more explicit if we use a custom event name.

Reply:
I'd be fine with `selection-change` if the event didn't bubble, otherwise we should namespace it.

Reply:
That is a really good point. Especially since we don't control what the slots contain. 

- [ ] Unique event name.